File: cpp/template/dfs.cpp
// Template 1. preorder DFS
// O(N) / O(H)
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        preorder(root, ans);
        return ans;
    }
private:
    void preorder(TreeNode* root, vector<int>& ans) {
        if (!root) return;
        ans.push_back(root->val);
        preorder(root->left, ans);
        preorder(root->right, ans);
    }
};


// Template 2. inorder DFS
// O(N) / O(H)
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> ans;
        inorder(root, ans);
        return ans;
    }
private:
    void inorder(TreeNode* node, vector<int>& ans) {
        if (!node) return;
        inorder(node->left, ans);
        ans.push_back(node->val);
        inorder(node->right, ans);
    }
};


// Template 3. postorder DFS
// O(N) / O(H)
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> ans;
        postorder(root, ans);
        return ans;
    }
private:
    void postorder(TreeNode* root, vector<int>& ans) {
        if (!root) return;
        postorder(root->left, ans);
        postorder(root->right, ans);
        ans.push_back(root->val);
    }
};

File: python/template/dfs.py
# Template 1. preorder DFS
# O(N) / O(H)
class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        arr = []
        def preorder(node):
            if not node:
                return
            arr.append(node.val)
            preorder(node.left)
            preorder(node.right)
        preorder(root)
        return arr


# Template 2. inorder DFS
# O(N) / O(H)
class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        arr = []
        def inorder(node):
            if not node:
                return
            inorder(node.left)
            arr.append(node.val)
            inorder(node.right)
        inorder(root)
        return arr


# Template 3. postorder DFS
# O(N) / O(H)
class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        arr = []
        def postorder(node):
            if not node:
                return
            postorder(node.left)
            postorder(node.right)
            arr.append(node.val)
        postorder(root)
        return arr